BinaryEncoding.md

```diff
@@ -3,120 +3,143 @@
 This document describes the [portable](Portability.md) binary encoding of the
 [Abstract Syntax Tree](AstSemantics.md) nodes.
 
-The binary encoding is designed to allow fast startup, which includes reducing
-download size and allow for quick decoding. For more information, see the 
-[rationale document](Rationale.md#why-a-binary-encoding)
-
-Reducing download size, is achieved through three layers:
-
- * The **raw** binary encoding itself, natively decoded by the browser, and to
-   be standardized in the [MVP](MVP.md).
- * **Specific** compression to the binary encoding, that is unreasonable to
-   expect a generic compression algorithm like gzip to achieve.
-   * This is not meant to be standardized, at least not initially, as it can be
```

Comment:
On Thu, Jan 21, 2016 at 8:25 PM, Alon Zakai <notifications@github.com>
wrote:

> In BinaryEncoding.md
> <https://github.com/WebAssembly/design/pull/520#discussion_r50449341>:
>
> > @@ -3,120 +3,143 @@
> >  This document describes the [portable](Portability.md) binary encoding of the
> >  [Abstract Syntax Tree](AstSemantics.md) nodes.
> >
> > -The binary encoding is designed to allow fast startup, which includes reducing
> > -download size and allow for quick decoding. For more information, see the
> > -[rationale document](Rationale.md#why-a-binary-encoding)
> > -
> > -Reducing download size, is achieved through three layers:
> > -
> > - * The **raw** binary encoding itself, natively decoded by the browser, and to
> > -   be standardized in the [MVP](MVP.md).
> > - * **Specific** compression to the binary encoding, that is unreasonable to
> > -   expect a generic compression algorithm like gzip to achieve.
> > -   * This is not meant to be standardized, at least not initially, as it can be
>
> We can get more real-world data now. I tested on zlib built with
> emscripten, and ran asm2wasm, then wasm-as which emits the current binary
> format. I see a 33% size reduction before gzip, and 16% size reduction
> after gzip.
>
I'm actually surprised by only 33% before compression. My fuzzy
recollection of the packer work from Luke was that it was around 3-5x
reduction, if not more. How big are the code sections compared to the
signature, function declaration, etc? Are you making use of the I8Const
shorter opcodes?



> And in addition to the 16% smaller download, we will have massively faster
> parsing. Overall that seems like a very compelling MVP to me.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/WebAssembly/design/pull/520/files#r50449341>.
>